File: python-sdk/examples/05_quote_logging_demo.py
#!/usr/bin/env python3
"""
Example 05: Quote Logging Demo

Demonstrates route discovery plus quote streaming with verbose SDK logging.
"""

import asyncio
import logging
import os

from kaleido_sdk import KaleidoClient, Layer

API_URL = os.getenv("KALEIDO_API_URL", "https://api.staging.kaleidoswap.com")
WS_URL = os.getenv("KALEIDO_WS_URL", "wss://api.staging.kaleidoswap.com/api/v1/market/ws")

# ---------------------------------------------------------------------------
# Logging setup (application's responsibility — the SDK never does this)
# ---------------------------------------------------------------------------
logging.basicConfig(
    level=logging.DEBUG,
    format="%(asctime)s [%(levelname)-8s] %(name)s — %(message)s",
    datefmt="%H:%M:%S",
)

# Keep useful SDK logs while cutting raw HTTP noise a bit.
logging.getLogger("kaleido_sdk.http").setLevel(logging.INFO)


async def main() -> None:
    """Main entry point."""
    client = KaleidoClient.create(
        base_url=API_URL,
        log_level=logging.DEBUG,
    )
    ws = client.maker.enable_websocket(WS_URL)

    print(f"Maker API: {API_URL}")
    print(f"WebSocket: {WS_URL}")
    print("-" * 40)

    print("\nDiscovering routes for BTC/USDT...")
    routes = await client.maker.get_available_routes("BTC", "USDT")
    if not routes:
        print("No routes found for BTC/USDT")
        return

    for index, route in enumerate(routes, start=1):
        print(f"  {index}. {route.from_layer} -> {route.to_layer}")

    quotes_received = 0
    finished = asyncio.Event()

    def on_quote(quote: dict) -> None:
        nonlocal quotes_received
        quotes_received += 1

        from_asset = quote.get("from_asset") or {}
        to_asset = quote.get("to_asset") or {}

        print("\n" + "=" * 50)
        print(f"Quote #{quotes_received}")
        print("=" * 50)
        print(f"RFQ ID: {quote.get('rfq_id')}")
        print(f"From: {from_asset.get('amount')} {from_asset.get('ticker')} ({from_asset.get('layer')})")
        print(f"To: {to_asset.get('amount')} {to_asset.get('ticker')} ({to_asset.get('layer')})")
        print(f"Price: {quote.get('price')}")

        fee = quote.get("fee") or {}
        print(f"Fee: {fee.get('final_fee')} {fee.get('fee_asset')}")
        print(f"Valid for: {quote.get('expires_at', 0) - quote.get('timestamp', 0)}s")

        if quotes_received >= 5:
            finished.set()

    print("\nStreaming quotes using the first discovered route...")
    stop = await client.maker.stream_quotes_by_ticker(
        from_ticker="BTC",
        to_ticker="USDT",
        amount=100_000,
        on_update=on_quote,
        preferred_from_layer=Layer(routes[0].from_layer),
        preferred_to_layer=Layer(routes[0].to_layer),
        poll_interval=2.0,
    )

    try:
        await asyncio.wait_for(finished.wait(), timeout=15.0)
    finally:
        stop()
        ws.disconnect()
        # disconnect() schedules the websocket close on the running loop
        await asyncio.sleep(0.1)
        await client.close()

    print(f"\nReceived {quotes_received} quotes. Done.")


if __name__ == "__main__":
    asyncio.run(main())

File: typescript-sdk/examples/05_quote_logging_demo.ts
#!/usr/bin/env tsx

/**
 * Example script to demonstrate quote logging with the new simplified API
 * Run this against a live maker to see detailed quote information
 */

import { KaleidoClient } from 'kaleido-sdk';

const API_URL = process.env.KALEIDO_API_URL || 'http://localhost:8000';
const API_KEY = process.env.KALEIDO_API_KEY;
const WS_URL = process.env.KALEIDO_WS_URL || 'ws://localhost:8000/api/v1/market/ws';

async function main() {
    console.log('🚀 Starting Quote Logging Demo\n');

    const client = await KaleidoClient.create({ baseUrl: API_URL, apiKey: API_KEY });

    try {
        // First, discover available routes
        console.log('🔍 Discovering available routes for BTC/USDT...\n');
        const routes = await client.maker.getAvailableRoutes('BTC', 'USDT');

        if (routes.length === 0) {
            console.log('❌ No routes found for BTC/USDT');
            return;
        }

        console.log(`✅ Found ${routes.length} available routes:`);
        routes.forEach((route, index) => {
            console.log(`   ${index + 1}. ${route.from_layer} → ${route.to_layer}`);
        });
        console.log('');

        // Enable WebSocket
        console.log(`📡 Connecting to WebSocket: ${WS_URL}\n`);
        client.maker.enableWebSocket(WS_URL);

        // Stream quotes using the new simplified API
        console.log('📊 Streaming quotes for BTC → USDT (using first available route)\n');

        const unsubscribe = await client.maker.streamQuotesByTicker(
            'BTC',
            'USDT',
            10000000, // 0.1 BTC
            (quote) => {
                console.log('\n' + '='.repeat(60));
                console.log('📊 QUOTE RECEIVED');
                console.log('='.repeat(60));
                console.log(`From Asset:    ${quote.from_asset.ticker}`);
                console.log(
                    `From Amount:   ${quote.from_asset.amount} (${(quote.from_asset.amount / 100000000).toFixed(8)} BTC)`,
                );
                console.log(`To Asset:      ${quote.to_asset.ticker}`);
                console.log(`To Amount:     ${quote.to_asset.amount}`);
                console.log(`Price:         ${quote.price}`);
                console.log(`Fee Asset:     ${quote.fee.fee_asset}`);
                console.log(`Fee Amount:    ${quote.fee.final_fee}`);
                console.log(`Fee Rate:      ${quote.fee.fee_rate}%`);
                console.log(`RFQ ID:        ${quote.rfq_id}`);
                console.log(`Timestamp:     ${new Date(quote.timestamp * 1000).toISOString()}`);
                console.log(`Expires At:    ${new Date(quote.expires_at * 1000).toISOString()}`);
                console.log(`Valid For:     ${quote.expires_at - quote.timestamp}s`);
                console.log('='.repeat(60) + '\n');
            },
        );

        // Wait for 10 seconds to receive quotes
        console.log('⏳ Listening for quotes (10 seconds)...\n');
        await new Promise((resolve) => setTimeout(resolve, 10000));

        // Unsubscribe
        console.log('\n✅ Unsubscribing from quotes\n');
        unsubscribe();
    } catch (error) {
        console.error('❌ Error:', error);
        process.exit(1);
    }
}

main().catch(console.error);
